### ISSUE TITLE
Network chaos unable to heal if previous container crashes

### ISSUE THREAD

**Issue by jaysunmah** [chaos-mesh/chaos-mesh#1884]
## Bug Report

**What version of Kubernetes are you using?**

```
$kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-16T00:04:31Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.14-airbnb0", GitCommit:"9931e7ec32e27c7ef21b3a734d8554a5ebadf00c", GitTreeState:"clean", BuildDate:"2020-12-02T19:00:47Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**What version of Chaos Mesh are you using?**

(We're using a docker image that was built and stored within my company's internal docker registry)
```
$kc exec -it chaos-controller-manager-754c4c56db-qkvvk -- /usr/local/bin/chaos-controller-manager -version
Controller manager Version: version.Info{GitVersion:"v0.9.0-156+a7b83630d43439", GitCommit:"a7b83630d43439b43c091842b10ed6ed6166f69b", BuildDate:"2021-01-12T19:01:35Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}
```
**What did you do?**

We're testing out the `NetworkChaos` resource on one of our pods. For simplicity, this pod has two containers, A and B. Container A is set to auto crash itself in the event of a network partition. When we inject this network partition, we can verify that A will begin crashing. As a result, when we try to heal this network partition, it appears from the chaos-daemon logs that since the previous container process no longer exists, it is unable to revert the iptable changes in both containers A and B. 

**What did you expect to see?**
I believe our desired behavior is so that the chaos-daemon is able to heal this network partition even if the original container crashes. We should not have to manually delete this pod to clear the iptable changes.

**What did you see instead?**
The network chaos is unable to be healed, and as a result the pod is corrupted. In order to "heal" this pod's network changes, we had to delete the pod and restart it.

**Output of chaosctl**

Output logs from chaos-daemon. As you can see, eventually the proc pid gets reset to 0, which I assume is because the previous process pid is no longer valid
```
2021-05-18T17:04:00.935Z	INFO	chaos-daemon-server	flush ipset	{"request": "ipsets:<name:\"test_src\" cidrs:\"172.21.129.109/32\" > ipsets:<name:\"test_tgt\" cidrs:\"0.0.0.0/1\" cidrs:\"128.0.0.0/2\" cidrs:\"192.0.0.0/3\" cidrs:\"224.0.0.0/4\" > container_id:\"docker://5e95968c41953c12a658a419df1dab1ec25b829f5c9ba7c1015410f717e84526\" "}
2021-05-18T17:04:00.936Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset create test_srcold hash:net"}
2021-05-18T17:04:00.936Z	INFO	chaos-daemon-server	create ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset create test_srcold hash:net"}
2021-05-18T17:04:00.937Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset add test_srcold 172.21.129.109/32"}
2021-05-18T17:04:00.937Z	INFO	chaos-daemon-server	add CIDR to ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset add test_srcold 172.21.129.109/32"}
2021-05-18T17:04:00.938Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset rename test_srcold test_src"}
2021-05-18T17:04:00.938Z	INFO	chaos-daemon-server	rename ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset rename test_srcold test_src"}
2021-05-18T17:04:00.939Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset swap test_srcold test_src"}
2021-05-18T17:04:00.939Z	INFO	chaos-daemon-server	swap ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset swap test_srcold test_src"}
2021-05-18T17:04:00.940Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset create test_tgtold hash:net"}
2021-05-18T17:04:00.940Z	INFO	chaos-daemon-server	create ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset create test_tgtold hash:net"}
2021-05-18T17:04:00.942Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 0.0.0.0/1"}
2021-05-18T17:04:00.942Z	INFO	chaos-daemon-server	add CIDR to ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 0.0.0.0/1"}
2021-05-18T17:04:00.943Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 128.0.0.0/2"}
2021-05-18T17:04:00.943Z	INFO	chaos-daemon-server	add CIDR to ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 128.0.0.0/2"}
2021-05-18T17:04:00.944Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 192.0.0.0/3"}
2021-05-18T17:04:00.944Z	INFO	chaos-daemon-server	add CIDR to ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 192.0.0.0/3"}
2021-05-18T17:04:00.945Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 224.0.0.0/4"}
2021-05-18T17:04:00.945Z	INFO	chaos-daemon-server	add CIDR to ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset add test_tgtold 224.0.0.0/4"}
2021-05-18T17:04:00.946Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset rename test_tgtold test_tgt"}
2021-05-18T17:04:00.946Z	INFO	chaos-daemon-server	rename ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset rename test_tgtold test_tgt"}
2021-05-18T17:04:00.947Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- ipset swap test_tgtold test_tgt"}
2021-05-18T17:04:00.947Z	INFO	chaos-daemon-server	swap ipset	{"command": "/usr/bin/nsenter -n/proc/16677/ns/net -- ipset swap test_tgtold test_tgt"}
2021-05-18T17:04:00.950Z	INFO	chaos-daemon-server	Set iptables chains	{"request": "chains:<name:\"OUTPUT/test_\" direction:OUTPUT ipsets:\"test_tgt\" target:\"DROP\" > chains:<name:\"INPUT/test_\" ipsets:\"test_tgt\" target:\"DROP\" > container_id:\"docker://5e95968c41953c12a658a419df1dab1ec25b829f5c9ba7c1015410f717e84526\" "}
2021-05-18T17:04:00.950Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -N CHAOS-INPUT"}
2021-05-18T17:04:00.952Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -F CHAOS-INPUT"}
2021-05-18T17:04:00.953Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S INPUT"}
2021-05-18T17:04:00.956Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -N CHAOS-OUTPUT"}
2021-05-18T17:04:00.958Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -F CHAOS-OUTPUT"}
2021-05-18T17:04:00.959Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S OUTPUT"}
2021-05-18T17:04:00.961Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -N OUTPUT/test_"}
2021-05-18T17:04:00.962Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -F OUTPUT/test_"}
2021-05-18T17:04:00.964Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S OUTPUT/test_"}
2021-05-18T17:04:00.965Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -A OUTPUT/test_ -m set --match-set test_tgt dst -j DROP -w 5"}
2021-05-18T17:04:00.967Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S CHAOS-OUTPUT"}
2021-05-18T17:04:00.968Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -A CHAOS-OUTPUT -j OUTPUT/test_"}
2021-05-18T17:04:00.970Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -N INPUT/test_"}
2021-05-18T17:04:00.972Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -F INPUT/test_"}
2021-05-18T17:04:00.973Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S INPUT/test_"}
2021-05-18T17:04:00.975Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -A INPUT/test_ -m set --match-set test_tgt src -j DROP -w 5"}
2021-05-18T17:04:00.977Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -S CHAOS-INPUT"}
2021-05-18T17:04:00.978Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- iptables -w -A CHAOS-INPUT -j INPUT/test_"}
2021-05-18T17:04:00.981Z	INFO	chaos-daemon-server	handling tc request	{"tcs": "container_id:\"docker://5e95968c41953c12a658a419df1dab1ec25b829f5c9ba7c1015410f717e84526\" "}
2021-05-18T17:04:00.982Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- tc qdisc del dev eth0 root"}
2021-05-18T17:04:00.983Z	INFO	chaos-daemon-server	adding prio	{"parent": 0}
2021-05-18T17:04:00.983Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- tc qdisc add dev eth0 root handle 1: prio bands 3 priomap 1 2 2 2 1 2 0 0 1 1 1 1 1 1 1 1"}
2021-05-18T17:04:00.985Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- tc qdisc add dev eth0 parent 1:1 handle 2: sfq"}
2021-05-18T17:04:00.987Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- tc qdisc add dev eth0 parent 1:2 handle 3: sfq"}
2021-05-18T17:04:00.988Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/16677/ns/net -- tc qdisc add dev eth0 parent 1:3 handle 4: sfq"}
2021-05-18T17:04:31.584Z	INFO	chaos-daemon-server	flush ipset	{"request": "container_id:\"docker://24ef249e5757fda611abc17c79faf967752ba823d4988cb97e4734d08d3887b6\" "}
2021-05-18T17:04:31.587Z	INFO	chaos-daemon-server	Set iptables chains	{"request": "container_id:\"docker://24ef249e5757fda611abc17c79faf967752ba823d4988cb97e4734d08d3887b6\" "}
2021-05-18T17:04:31.587Z	INFO	background-process-manager	build command	{"command": "nsenter -n/proc/0/ns/net -- iptables -w -N CHAOS-INPUT"}
2021-05-18T17:04:31.588Z	ERROR	chaos-daemon-server	error while initializing iptables	{"error": "error code: exit status 1, msg: nsenter: cannot open /proc/0/ns/net: No such file or directory\n"}
```

**Comment by jaysunmah** [comment]
As a follow up question to this issue, can someone clarify the exact behavior of how the chaos-daemon injects these iptable changes onto the containers of a pod? From my understanding, running `nsenter ... ` commands should only affect one container in a pod. However, from the logs it looks like the chaos-daemon only tries to run this command once on a given pod, but it looks like all the containers in a pod are affected by these iptable changes

**Comment by YangKeao** [comment]
> As a follow up question to this issue, can someone clarify the exact behavior of how the chaos-daemon injects these iptable changes onto the containers of a pod? From my understanding, running `nsenter ... ` commands should only affect one container in a pod. However, from the logs it looks like the chaos-daemon only tries to run this command once on a given pod, but it looks like all the containers in a pod are affected by these iptable changes

Both containers are affected because they share the same network namespace, and all network operations in this network namespace will affect all of these processes.

**Comment by jaysunmah** [comment]
> > As a follow up question to this issue, can someone clarify the exact behavior of how the chaos-daemon injects these iptable changes onto the containers of a pod? From my understanding, running `nsenter ... ` commands should only affect one container in a pod. However, from the logs it looks like the chaos-daemon only tries to run this command once on a given pod, but it looks like all the containers in a pod are affected by these iptable changes
> 
> Both containers are affected because they share the same network namespace, and all network operations in this network namespace will affect all of these processes.

@YangKeao thanks for the insight! so a command such as `nsenter -n/proc/16677/ns/net -- iptables -w -A CHAOS-OUTPUT -j OUTPUT/test_` will affect process pid `16677`'s iptables, along with all the other pod's container process pids that share the same network namespace?

**Comment by YangKeao** [comment]
> along with all the other pod's container process pids that share the same network namespace?

Yes :beers:

**Comment by jaysunmah** [comment]
hey @YangKeao , bumping this thread. Would you or someone else at chaos-mesh org be able to address the initial problem, which is that I'm unable to cleanup a network chaos if the original container has crashed? Specifically, I believe this is the exact [line of code](https://github.com/chaos-mesh/chaos-mesh/blob/master/pkg/chaosdaemon/ipset_server.go#L38) that I'm erroring at, and ideally I'd like it to either
1) retry fetching a new containerID from the affected pod to cleanup the chaos, or
2) try to use another container and its containerID from the affected pod to cleanup the chaos (this option would technically work in our cases, but I feel like option 1 is better in the long run)

**Comment by jaysunmah** [comment]
I can also help out if needed, I'm having trouble connecting the point of code where this request's containerID is sent to `chaos-daemon-server` in the first place. For example, would you be able to give me a high level walkthrough of what happens when I run `kubectl delete -f chaosnetwork_test.yaml` ? It seems like somehow the containerID is store somewhere in memory with respect to the chaos name, so when I try deleting the chaos it'll automatically use this containerID instead of trying to fetch it again.

**Comment by jaysunmah** [comment]
Here's my error log stacktrace, with an additional info log that I added to try and debug:
```
2021-05-26T17:03:37.167Z	INFO	chaos-daemon-server	flush ipset	{"request": "container_id:\"docker://098c665fc5f5758aebd73639b51d5b7c54b8f8bafa2e44024e6ad21e9a2a901e\" enterNS:true "}
2021-05-26T17:03:37.167Z	INFO	chaos-daemon-server	DockerClient get pid from container id	{"containerID": "docker://098c665fc5f5758aebd73639b51d5b7c54b8f8bafa2e44024e6ad21e9a2a901e"}
2021-05-26T17:03:37.168Z	ERROR	chaos-daemon-server	error while getting PID	{"error": "container is not running, status: exited"}
github.com/go-logr/zapr.(*zapLogger).Error
	/tmp/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon.(*DaemonServer).FlushIPSets
	/mnt/pkg/chaosdaemon/ipset_server.go:38
github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon/pb._ChaosDaemon_FlushIPSets_Handler.func1
	/mnt/pkg/chaosdaemon/pb/chaosdaemon.pb.go:1917
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1
	/tmp/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/tmp/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.0/chain.go:25
github.com/chaos-mesh/chaos-mesh/pkg/grpc.TimeoutServerInterceptor
	/mnt/pkg/grpc/utils.go:100
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/tmp/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.0/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
	/tmp/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.0/chain.go:34
github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon/pb._ChaosDaemon_FlushIPSets_Handler
	/mnt/pkg/chaosdaemon/pb/chaosdaemon.pb.go:1919
google.golang.org/grpc.(*Server).processUnaryRPC
	/tmp/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024
google.golang.org/grpc.(*Server).handleStream
	/tmp/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313
google.golang.org/grpc.(*Server).serveStreams.func1.1
	/tmp/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722
```

**Comment by AsterNighT** [comment]
> I can also help out if needed, I'm having trouble connecting the point of code where this request's containerID is sent to `chaos-daemon-server` in the first place. For example, would you be able to give me a high level walkthrough of what happens when I run `kubectl delete -f chaosnetwork_test.yaml` ? It seems like somehow the containerID is store somewhere in memory with respect to the chaos name, so when I try deleting the chaos it'll automatically use this containerID instead of trying to fetch it again.

You can, in fact, delete the chaos by mannually removing the finalizer of the chaos. It is a long known bug due to it tries to recover in a deleted container. This is to be fixed in 2.0 release.

**Comment by jaysunmah** [comment]
@AsterNighT thanks for the suggestion. I was in fact deleting my networkchaos this way for the time being, but I don't believe this actually heals the underlying chaos. I would still need to manually delete and restart the pod, which is something me and my team would prefer to not have to do .
